crates/compiler/mono/src/reset_reuse.rs

```diff
@@ -398,49 +404,106 @@ fn insert_reset_reuse_operations_stmt<'a, 'i>(
-                {
+            enum SymbolIsUnique {
+                Never,
+                Always(Symbol),
+                MustCheck(Symbol),
+            }
+
+            let can_reuse = match rc {
+                ModifyRc::Dec(symbol) | ModifyRc::DecRef(symbol) => {
+                    // can only reuse if the symbol is (potentially) unique
+                    if environment.non_unique_symbols.contains(symbol) {
+                        SymbolIsUnique::Never
+                    } else {
+                        SymbolIsUnique::MustCheck(*symbol)
+                    }
+                }
+                ModifyRc::Free(symbol) => {
+                    // a free'd symbol is guaranteed to be unique
+                    SymbolIsUnique::Always(*symbol)
+                }
+                ModifyRc::Inc(_, _) => {
+                    // an incremented symbol is never unique
+                    SymbolIsUnique::Never
+                }
+            };
+
+            enum ResetOperation {
+                Reset,
+                ResetRef,
+                ClearTagId,
+                Nothing,
+            }
+
+            let reuse_pair = match can_reuse {
+                SymbolIsUnique::MustCheck(symbol) | SymbolIsUnique::Always(symbol) => {
                     // Get the layout of the symbol from where it is defined.
-                    let layout_option = environment.get_symbol_layout(*symbol);
+                    let layout_option = environment.get_symbol_layout(symbol);
 
                     // If the symbol is defined in the current proc, we can use the layout from the environment.
-                    match layout_option.clone() {
+                    match layout_option {
                         LayoutOption::Layout(layout) => {
                             match symbol_layout_reusability(
                                 layout_interner,
                                 environment,
-                                symbol,
+                                &symbol,
                                 layout,
                             ) {
                                 Reuse::Reusable(union_layout) => {
+                                    let (reuse_symbol, reset_op) = match rc {
+                                        ModifyRc::Dec(_) => (
+                                            Symbol::new(home, ident_ids.gen_unique()),
+                                            ResetOperation::Reset,
+                                        ),
+                                        ModifyRc::DecRef(_) => (
+                                            Symbol::new(home, ident_ids.gen_unique()),
+                                            ResetOperation::ResetRef,
+                                        ),
+                                        ModifyRc::Free(_) => {
+                                            if union_layout
+                                                .stores_tag_id_in_pointer(environment.target_info)
```

Comment:
Why does this require target info?

Reply:
the number of available bits changes. on 32-bit systems we can store 4 tags in a pointer, on 64-bit systems it is 8. So depending on the target, a 5-element tag union may or may not store the tag in the pointer.

Reply:
Could you update the `get_reuse_layout_info` function in `reset_reuse`? I think should take the same info into account.